feat(agility): Add efficient alching mode with improved obstacle completion detection

runsonmypc opened the pull request:
This PR introduces an efficient alching mode for the agility plugin along with improved obstacle completion detection to prevent mid-obstacle interactions, and introduces XP Drops as an early signal for obstacle completion.

## Features Added:

- **Efficient Alching Mode**: New config option that changes alching behavior for obstacles 5+ tiles away
  - When enabled: Click obstacle first → Alch during movement → Click obstacle again
  - Maximizes efficiency by utilizing movement time for alching
- **Skip Inefficient Option**: New config to only alch when obstacles are 5+ tiles away
  - Prevents suboptimal alching on close obstacles to maintain maximum XP rates
- **XP Drop Detection**: Uses agility XP drops to instantly detect obstacle completion
  - Allows immediate action on next obstacle without waiting for animations to finish
  - Significantly faster reactions compared to movement-based detection alone for some obstacles
- **Alch Skip Chance**: Configurable percentage chance (0-100%) to randomly skip alching on any obstacle
  - Applies to all obstacles and all alching modes

## Bug Fixes:

- **Mid-Obstacle Interaction Prevention**: Added XP-based completion detection and timing delays to prevent mid-obstacle clicking during animation pauses (e.g., obstacles that have a small pause before the character jumps)
- **Colossal Wyrm Support**: Disabled XP checks for Colossal Wyrm courses which have obstacles with multiple XP drops
- **Rs2Walker Alching Fix**: Prevents alching immediately after Rs2Walker returns to course start

## Technical Improvements:

- Added XP tracking to detect obstacle completion immediately upon XP gain for faster XP rates
- Special handling for first obstacle (index 0) to prevent issues after lap completion
- **Refactored obstacle completion detection** into course-specific `isObstacleComplete()` method
- **Reduced code nesting** by extracting alching logic into dedicated helper methods
- **Improved code maintainability** through better separation of concerns

## Testing Notes:

Please let me know if you know of any other courses that have obstacles with XP drops before they end

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Added configurable alching options: efficient alching, skip inefficient alchs (nearby obstacles), and per-obstacle skip chance (0–100%).
  - Integrated alching into the agility loop with distance-aware and randomized skip behavior.

- Refactor
  - Introduced XP-aware gating for obstacle clicks to avoid premature interactions.
  - Replaced blocking waits with polling timeouts and improved movement/animation/plane/health detection.
  - Consolidated course-specific handling for more reliable runs across courses (Colossal Wyrm, Prifddinas, Werewolf).
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
feat(agility): add efficient alching mode

- Add new config option for efficient alching
- When enabled and obstacle is 5+ tiles away, clicks obstacle first, then alchs during movement, then clicks again
- Add timestamp-based waiting to prevent mid-obstacle interactions during animation pauses
- Track agility XP to detect obstacle completion
- Fix double alching bug on first obstacle after lap completion

Commit:
fix(agility): randomize animation pause detection delay

- Change hardcoded 1000ms delay to random 700-1100ms range
- Makes bot behavior less predictable and more human-like
- Applies to both Colossal Wyrm XP check disable and general pause detection

Commit:
feat(agility): add humanization delay after XP drop detection

- Add 100-300ms random delay after detecting XP gain
- Makes reactions more human-like while still faster than timeout

Commit:
feat(agility): add configurable alch skip chance

- Add config option to randomly skip alching with 0-100% chance
- Default set to 5% skip chance for more human-like behavior
- Applies to all obstacles including first obstacle (index 0)
- Works with both normal and efficient alching modes

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/agility/AgilityScript.java:
```diff
@@ -96,10 +101,24 @@ public boolean run()
-					if (Rs2Player.isMoving() || Rs2Player.isAnimating())
-					{
+					// Disable XP checks for Colossal Wyrm courses (they have multi-XP drop obstacles)
+					boolean isColossalWyrm = config.agilityCourse().name().contains("COLOSSAL_WYRM");
```
I think the real issue here is the waitForCompletion method from the CourseHandler interface.

This method should be holding this future until the obstacle is completed.


We can create exceptions for specfic courses by overriding this method in the course's class, or globally change it by changing it in the interface.

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/agility/AgilityScript.java:
```diff
@@ -170,11 +184,69 @@ else if (!(plugin.getCourseHandler() instanceof GnomeStrongholdCourse))
-				{
-					plugin.getCourseHandler().waitForCompletion(agilityExp, Microbot.getClient().getLocalPlayer().getWorldLocation().getPlane());
+				// Handle alchemy if enabled
+				if (config.alchemy()) {
```
Can this be refactored to not be so, nested? I think we could create a method on the interface & override it if there is going to be exceptions for specfic courses.

Commit:
refactor(agility): improve obstacle completion detection and reduce code nesting

  - Add isObstacleComplete() method to AgilityCourseHandler for course-specific completion logic
  - Override completion detection in Colossal Wyrm courses to handle multi-XP drops
  - Extract alching logic into separate helper methods (shouldPerformAlch, performEfficientAlch, performNormalAlch)
  - Consolidate course-specific actions into handleCourseSpecificActions()
  - Reduce nesting and improve code readability per PR feedback

Commit:
fix(agility): prevent alching immediately after Rs2Walker returns to start

  - Add Rs2Player.isWalking() check at obstacle 0 to skip alching while walking
  - Ensures obstacle is clicked first after returning to course start

Commit:
feat(agility): add skip inefficient alching option

  - Add config to skip alching when obstacle is <5 tiles away
  - Move skip inefficient config above alch skip chance
  - Fix return handling for efficient alch in skip inefficient mode

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/agility/AgilityScript.java:
```diff
@@ -97,10 +102,26 @@ public boolean run()
 				if (plugin.getCourseHandler().getCurrentObstacleIndex() > 0)
 				{
-					if (Rs2Player.isMoving() || Rs2Player.isAnimating())
+					boolean obstacleComplete = plugin.getCourseHandler().isObstacleComplete(
```
Shouldn't the functionality be at the end of the loop? seems like you wanted to replace waitForCompletion.

runsonmypc replied:
@g-mason0 Thanks for the feedback! You're right - the logic was scattered. I've refactored to centralize everything in the course handler:

- Added shouldClickObstacle() method that handles the "don't click while animating unless XP gained" logic
- Kept waitForCompletion() for post-click blocking
- Removed duplicate XP tracking and animation checks from the main script

Now the handler controls all timing decisions. The script just asks "should I click?" and the handler decides based on animation/XP state. Much cleaner than having these checks spread across multiple places.

The XP drop detection for early action still works - if we get XP while animating, shouldClickObstacle() returns true and we can click the next obstacle immediately.

Commit:
fix(agility): (WIP) fix XP drop detection logic placement

  - Keep waitForCompletion() as blocking post-click check (not replaced)
  - Add pre-click animation guard to prevent clicks while animating without XP
  - Move animation check before alching code to guard all click paths
  - Remove redundant obstacle completion check that was causing conflicts

  Addresses g-mason0's comment: The XP detection is intentionally split -
  pre-click check prevents premature clicks, post-click waitForCompletion
  blocks until XP/timeout. This ensures we only act while animating when
  XP drop signals completion.

Commit:
fix(agility): centralize XP detection logic in AgilityCourseHandler

  - Move animation/XP checking to shouldClickObstacle() method in handler
  - Remove duplicate XP tracking update after waitForCompletion()
  - Simplify flow: only update lastAgilityXp when XP changes (line 137)

  Addresses g-mason0's comment: XP detection logic now lives in the course
  handler methods rather than scattered in the script. The handler determines
  when to click (shouldClickObstacle) and when complete (waitForCompletion).

Commit:
fix(agility): disable XP-based early clicking for Colossal Wyrm courses

  - Override shouldClickObstacle() in both Wyrm courses to ignore XP drops
  - Some Wyrm obstacles have multiple XP drops per animation, making early action unreliable
  - Maintains standard animation-based timing for these specific courses

Commit:
fix(agility): add input validation for alchSkipChance config

- Add @Range(min = 0, max = 100) annotation to validate percentage bounds
- Prevents misconfiguration with values outside 0-100 range

Commit:
fix(agility): optimize alch flow and reduce log spam

- Remove duplicate getAlchItem() call from shouldPerformAlch()
- Comment out log statements that spam at 10 Hz when misconfigured
- Inventory now scanned once per tick instead of twice

Commit:
fix(agility): handle waitForCompletion timeout to avoid stale state actions

Commit:
refactor(agility): remove unused currentAgilityXp parameter from performNormalAlch

Commit:
fix(agility): integrate isObstacleComplete hook for course-specific completion logic

runsonmypc commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/agility/courses/ColossalWyrmBasicCourse.java:
```diff
@@ -1,35 +1,100 @@
+		
+		// Timeout reached
+		return false;
+	}
```
Further changes to other courses are out of scope.

Commit:
fix(agility): normalize line endings to LF

gmason0 reviewed (approved):
there is still some heavy nesting here, but otherwise this seems to be alot better than before.

We can move forward with this - thank for your being on top of this despite the long review period.
